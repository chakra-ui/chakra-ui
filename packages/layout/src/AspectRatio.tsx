import { __DEV__ } from "@chakra-ui/utils"
import React, { forwardRef, Children } from "react"
import { Box, BoxProps } from "./Box"

interface AspectRatioOptions {
  /**
   * The aspect ratio of the Box. Common values are:
   *
   * `21/9`, `16/9`, `9/16`, `4/3`, `1.85/1`
   */
  ratio?: number
}

export type AspectRatioProps = BoxProps & AspectRatioOptions

/**
 * AspectRatio
 *
 * Mostly useful for cropping media (videos, images and maps) to a desired aspect ratio.
 *
 * @see Docs https://chakra-ui.com/aspectratio
 */
export const AspectRatio = forwardRef(
  (props: AspectRatioProps, ref: React.Ref<any>) => {
    const { ratio = 4 / 3, children, ...rest } = props
    const child = Children.only(children)

    return (
      <Box
        ref={ref}
        position="relative"
        _before={{
          height: 0,
          content: `""`,
          display: "block",
          paddingBottom: `${(1 / ratio) * 100}%`,
        }}
        css={{
          "& > *": {
            overflow: "hidden",
            position: "absolute",
            top: "0",
            right: "0",
            bottom: "0",
            left: "0",
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
            width: "100%",
            height: "100%",
          },
          "& > img, & > video": {
            objectFit: "cover",
          },
        }}
        {...rest}
      >
        {child}
      </Box>
    )
  },
)

if (__DEV__) {
  AspectRatio.displayName = "AspectRatio"
}
