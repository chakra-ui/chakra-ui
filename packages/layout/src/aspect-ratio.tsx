import {
  chakra,
  forwardRef,
  ResponsiveValue,
  HTMLChakraProps,
} from "@chakra-ui/system"
import { cx, mapResponsive, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

interface AspectRatioOptions {
  /**
   * The aspect ratio of the Box. Common values are:
   *
   * `21/9`, `16/9`, `9/16`, `4/3`, `1.85/1`
   */
  ratio?: ResponsiveValue<number>
}

export interface AspectRatioProps
  extends HTMLChakraProps<"div">,
    AspectRatioOptions {}

/**
 * React component used to cropping media (videos, images and maps)
 * to a desired aspect ratio.
 *
 * @see Docs https://chakra-ui.com/docs/layout/aspect-ratio
 */
export const AspectRatio = forwardRef<AspectRatioProps, "div">((props, ref) => {
  const { ratio = 4 / 3, children, className, ...rest } = props

  // enforce single child
  const child = React.Children.only(children)

  const _className = cx("chakra-aspect-ratio", className)

  return (
    <chakra.div
      ref={ref}
      position="relative"
      className={_className}
      _before={{
        height: 0,
        content: `""`,
        display: "block",
        paddingBottom: mapResponsive(ratio, (r) => `${(1 / r) * 100}%`),
      }}
      __css={{
        "& > *:not(style)": {
          overflow: "hidden",
          position: "absolute",
          top: "0",
          right: "0",
          bottom: "0",
          left: "0",
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
          width: "100%",
          height: "100%",
        },
        "& > img, & > video": {
          objectFit: "cover",
        }
      }}
      {...rest}
    >
      {child}
    </chakra.div>
  )
})

if (__DEV__) {
  AspectRatio.displayName = "AspectRatio"
}
