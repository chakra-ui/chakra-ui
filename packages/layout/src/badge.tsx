import * as React from "react"
import {
  chakra,
  PropsOf,
  useStyleConfig,
  omitThemingProps,
  ThemingProps,
} from "@chakra-ui/system"
import { __DEV__, cx } from "@chakra-ui/utils"

export type BadgeProps = PropsOf<typeof chakra.span> & ThemingProps

/**
 * React component used to display notifications, messages, or
 * statuses in different shapes and sizes.
 *
 * @see Docs https://chakra-ui.com/components/badge
 */
export const Badge = React.forwardRef(function Badge(
  props: BadgeProps,
  ref: React.Ref<any>,
) {
  const styles = useStyleConfig("Badge", props)
  const { className, ...rest } = omitThemingProps(props)

  return (
    <chakra.span
      ref={ref}
      className={cx("chakra-badge", props.className)}
      {...rest}
      __css={{
        display: "inline-block",
        whiteSpace: "nowrap",
        verticalAlign: "middle",
        ...styles,
      }}
    />
  )
})

if (__DEV__) {
  Badge.displayName = "Badge"
}
