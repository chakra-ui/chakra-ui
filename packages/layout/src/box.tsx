import {
  chakra,
  PropsOf,
  forwardRef,
  SystemStyleObject,
} from "@chakra-ui/system"
import { __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

export type BoxProps = PropsOf<typeof Box>

/**
 * Box is the most abstract component on top of which other chakra
 * components are built. It renders a `div` element by default.
 *
 * @see Docs https://chakra-ui.com/docs/layout/box
 */
export const Box = chakra.div

if (__DEV__) {
  Box.displayName = "Box"
}

/**
 * As a constraint, you can't pass size related props
 * Only `size` would be allowed
 */
type Omitted = "size" | "boxSize" | "width" | "height" | "w" | "h"

export type SquareProps = Omit<BoxProps, Omitted> & {
  /**
   * The size (width and height) of the square
   */
  size?: BoxProps["width"]
  /**
   * If `true`, the content will be centered in the square
   */
  centerContent?: boolean
}

export const Square = forwardRef<SquareProps>(function Square(props, ref) {
  const { size, centerContent = true, ...rest } = props
  const centerStyles: SystemStyleObject = centerContent
    ? { display: "flex", alignItems: "center", justifyContent: "center" }
    : {}

  return (
    <Box
      ref={ref}
      boxSize={size}
      __css={{
        ...centerStyles,
        flexShrink: 0,
        flexGrow: 0,
      }}
      {...rest}
    />
  )
})

if (__DEV__) {
  Square.displayName = "Square"
}

export const Circle = forwardRef<SquareProps>(function Circle(props, ref) {
  const { size, ...rest } = props
  return <Square size={size as any} ref={ref} borderRadius="9999px" {...rest} />
})

if (__DEV__) {
  Circle.displayName = "Circle"
}
