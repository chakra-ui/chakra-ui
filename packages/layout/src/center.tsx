import { chakra, forwardRef, HTMLChakraProps } from "@chakra-ui/system"
import { __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

export interface CenterProps extends HTMLChakraProps<"div"> {}

/**
 * React component used to horizontally and vertically center its child.
 * It uses the popular `display: flex` centering technique.
 *
 * @see Docs https://chakra-ui.com/center
 */
export const Center = chakra("div", {
  baseStyle: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
})

if (__DEV__) {
  Center.displayName = "Center"
}

export interface AbsoluteCenterProps extends HTMLChakraProps<"div"> {
  axis?: "horizontal" | "vertical" | "both"
}

const centerStyles = {
  horizontal: {
    insetStart: "50%",
    transform: "translateX(-50%)",
  },
  vertical: {
    top: "50%",
    transform: "translateY(-50%)",
  },
  both: {
    insetStart: "50%",
    top: "50%",
    transform: "translate(-50%, -50%)",
  },
}

/**
 * React component used to horizontally and vertically center an element
 * relative to its parent dimensions.
 *
 * It uses the `position: absolute` strategy.
 *
 * @see Docs https://chakra-ui.com/center
 * @see WebDev https://web.dev/centering-in-css/#5.-pop-and-plop
 */
export const AbsoluteCenter = forwardRef<AbsoluteCenterProps, "div">(
  (props, ref) => {
    const { axis = "both", ...rest } = props
    return (
      <chakra.div
        ref={ref}
        __css={centerStyles[axis]}
        {...rest}
        position="absolute"
      />
    )
  },
)
