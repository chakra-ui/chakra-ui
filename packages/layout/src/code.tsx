import {
  chakra,
  forwardRef,
  omitThemingProps,
  ThemingProps,
  useStyleConfig,
  HTMLChakraProps,
} from "@chakra-ui/system"
import { cx, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"

export interface CodeProps
  extends HTMLChakraProps<"code">,
    ThemingProps<"Code"> {}

/**
 * React component to render inline code snippets.
 *
 * @see Docs https://chakra-ui.com/code
 */
export const Code = forwardRef<CodeProps, "code">((props, ref) => {
  const styles = useStyleConfig("Code", props)
  const { className, ...rest } = omitThemingProps(props)

  return (
    <chakra.code
      ref={ref}
      className={cx("chakra-code", props.className)}
      {...rest}
      __css={{
        display: "inline-block",
        ...styles,
      }}
    />
  )
})

if (__DEV__) {
  Code.displayName = "Code"
}
