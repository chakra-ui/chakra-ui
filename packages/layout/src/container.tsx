import { chakra, PropsOf, useTheme, SystemStyleObject } from "@chakra-ui/system"
import {
  cx,
  Dict,
  get,
  mapResponsive,
  __DEV__,
  filterUndefined,
} from "@chakra-ui/utils"
import * as React from "react"

export type ContainerProps = PropsOf<typeof chakra.div> & {
  /**
   * If `true`, container will center it's children
   * regardless of their width.
   */
  centerContent?: boolean
}

/**
 * Layout component used to wrap app or website content
 *
 * It sets `margin-left` and `margin-right` to `auto`,
 * to keep it's content centered.
 *
 * It also sets a default max-width of `60ch` (60 characters).
 */
export const Container = React.forwardRef(function Container(
  props: ContainerProps,
  ref: React.Ref<any>,
) {
  const {
    maxWidth,
    width,
    minWidth,
    w,
    minW,
    maxW,
    className,
    centerContent,
    ...rest
  } = props

  const theme = useTheme()

  const widthProps = transform(theme, {
    maxW,
    maxWidth,
    width,
    w,
    minWidth,
    minW,
  })

  const styles: SystemStyleObject = {
    w: "100%",
    mx: "auto",
    maxW: "60ch",
    px: "1rem",
    ...(centerContent && {
      display: "flex",
      flexDirection: "column",
      alignItems: "center",
    }),
    ...widthProps,
  }

  return (
    <chakra.div
      ref={ref}
      className={cx("chakra-container", className)}
      {...rest}
      __css={styles}
    />
  )
})

if (__DEV__) {
  Container.displayName = "Container"
}

function transform(theme: Dict, props: Dict) {
  const result: SystemStyleObject = {}

  for (const prop in props) {
    const propValue = props[prop]
    result[prop] = mapResponsive(propValue, (value) =>
      get(theme, `sizes.container.${value}`, value),
    )
  }

  return filterUndefined(result)
}
