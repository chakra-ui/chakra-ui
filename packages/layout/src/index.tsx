export * from "./aspect-ratio"
export * from "./badge"
export * from "./box"
export * from "./center"
export * from "./code"
export * from "./container"
export * from "./divider"
export * from "./flex"
export * from "./grid"
export * from "./heading"
export * from "./kbd"
export * from "./link"
export * from "./list"
export * from "./simple-grid"
export * from "./spacer"
export * from "./stack"
export * from "./text"
export * from "./wrap"
export * from "./link-box"
