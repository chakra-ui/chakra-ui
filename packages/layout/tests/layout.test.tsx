import { render, testA11y } from "@chakra-ui/test-utils"
import * as React from "react"
import { ChakraProvider, extendTheme } from "@chakra-ui/react"
import { Box, Badge, Container, Divider, Flex } from "../src"

describe("<Box />", () => {
  test("passes a11y test", async () => {
    await testA11y(<Box>this is a box</Box>)
  })

  test("as - prop works correctly", () => {
    const { getByText } = render(
      <Box as="a" href="www.google.com">
        Box
      </Box>,
    )
    expect(getByText("Box").nodeName).toBe("A")
  })
})

describe("<Badge />", () => {
  test("passes a11y test", async () => {
    await testA11y(<Badge>this is a badge</Badge>)
  })
})

describe("<Container />", () => {
  test("renders box correctly", () => {
    render(<Container>This is container</Container>)
  })

  test("centerContent - prop works correctly", () => {
    render(<Container centerContent>This is centered container</Container>)
  })

  test("themeing works correctly", () => {
    const theme = extendTheme({
      components: {
        Container: {
          variants: {
            customBackground: {
              bgColor: "red.500",
            },
          },
        },
      },
    })
    render(
      <ChakraProvider theme={theme}>
        <Container variant="customBackground">
          This is container has a red background
        </Container>
      </ChakraProvider>,
    )
  })
})

describe("<Flex />", () => {
  test("renders all the allowed shorthand style props", () => {
    render(
      <Flex
        align="stretch"
        justify="start"
        wrap="nowrap"
        direction="row"
        basis="auto"
        grow={1}
        shrink={0}
      />,
    )
  })
})

describe("<Divider />", () => {
  test("renders with default theming", () => {
    render(<Divider />)
  })

  test("overrides the theming props", () => {
    render(<Divider variant="dashed" />)
  })
})
