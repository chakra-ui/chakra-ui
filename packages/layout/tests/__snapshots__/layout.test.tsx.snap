// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge override the theming props 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  white-space: nowrap;
  vertical-align: middle;
  padding-left: 0.25rem;
  padding-right: 0.25rem;
  text-transform: uppercase;
  font-size: 0.75rem;
  border-radius: 0.125rem;
  font-weight: 700;
  color: #D53F8C;
  box-shadow: inset 0 0 0px 1px #D53F8C;
}

<span
    class="chakra-badge emotion-0"
  >
    Badge
  </span>
</DocumentFragment>
`;

exports[`Badge should render with default theming 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  white-space: nowrap;
  vertical-align: middle;
  padding-left: 0.25rem;
  padding-right: 0.25rem;
  text-transform: uppercase;
  font-size: 0.75rem;
  border-radius: 0.125rem;
  font-weight: 700;
  background: #EDF2F7;
  color: #1A202C;
}

<span
    class="chakra-badge emotion-0"
  >
    Badge
  </span>
</DocumentFragment>
`;

exports[`Box as - prop works correctly 1`] = `
<DocumentFragment>
  <a
    class="css-0"
    href="www.google.com"
  >
    This is box
  </a>
</DocumentFragment>
`;

exports[`Box should render box correctly 1`] = `
<DocumentFragment>
  <div
    class="css-0"
  >
    This is box
  </div>
</DocumentFragment>
`;
